pkg/register/register.go

```diff
@@ -76,53 +116,60 @@ func Register(url string, caCert []byte, smbios bool, emulatedTPM bool, emulated
 		}
 		return nil, err
 	}
-	defer conn.Close()
 	_ = conn.SetWriteDeadline(time.Now().Add(RegistrationDeadlineSeconds * time.Second))
 	_ = conn.SetReadDeadline(time.Now().Add(RegistrationDeadlineSeconds * time.Second))
 
-	logrus.Debug("start TPM attestation")
-	err = tpmAuth.Init(conn)
-	if err != nil {
-		return nil, fmt.Errorf("failed TPM based authentication: %w", err)
+	return conn, nil
+}
+
+func doTPMAttestation(tpmAuth *tpm.AuthClient, conn *websocket.Conn) error {
+	if err := tpmAuth.Init(conn); err != nil {
+		return err
 	}
 
 	msgType, _, err := ReadMessage(conn)
 	if err != nil {
-		return nil, fmt.Errorf("expecting auth reply: %w", err)
+		return fmt.Errorf("expecting auth reply: %w", err)
 	}
 	if msgType != MsgReady {
-		return nil, fmt.Errorf("expecting '%v' but got '%v'", MsgReady, msgType)
+		return fmt.Errorf("expecting '%v' but got '%v'", MsgReady, msgType)
 	}
-	logrus.Info("TPM attestation successful")
+	return nil
+}
 
-	err = sendData(conn, smbios)
-	if err != nil {
-		return nil, err
+func negotiateProtoVersion(conn *websocket.Conn) (MessageType, error) {
```

Comment:
pretty cool! This could allow us to even maintain backwards compat much more easily if needed!